src/inmanta/module.py

```diff
@@ -276,7 +320,172 @@ def get_file_for_version(self, repo: str, tag: str, file: str) -> str:
 gitprovider = CLIGitProvider()
 
 
-class ModuleRepo(object):
+# TODO: test all relevant module source methods, including editable install detection
+class ModuleSource(Generic[TModule]):
+    def get_module(
+        self, project: "Project", module_spec: List[InmantaModuleRequirement], install: bool = False
+    ) -> Optional[TModule]:
+        """
+        Returns the appropriate module instance for a given module spec.
+
+        :param project: The project associated with the module.
+        :param module_spec: The module specification including any constraints on its version. Ignored if module
+            is already installed. In this case, the project is responsible for verifying constraint compatibility.
+        :param install: Whether to attempt to install the module if it hasn't been installed yet.
+        """
+        module_name: str = self._get_module_name(module_spec)
+        path: Optional[str] = self.path_for(module_name)
+        if path is not None:
+            return self.from_path(project, path)
+        elif install:
+            return self.install(project, module_spec)
+        else:
+            return None
+
+    @abstractmethod
+    def install(self, project: "Project", module_spec: List[InmantaModuleRequirement]) -> Optional[TModule]:
+        """
+        Attempt to install a module given a module spec.
+
+        :param project: The project associated with the module.
+        :param module_spec: The module specification including any constraints on its version.
+        """
+        raise NotImplementedError("Abstract method")
+
+    def path_for(self, name: str) -> Optional[str]:
+        """
+        Returns the path to the module root directory. Should be called prior to configuring the module finder for v1 modules.
+        """
+        raise NotImplementedError("Abstract method")
+
+    @classmethod
+    @abstractmethod
+    def from_path(cls, project: Optional["Project"], path: str) -> TModule:
+        raise NotImplementedError("Abstract method")
+
+    def _get_module_name(self, module_spec: List[InmantaModuleRequirement]) -> str:
+        module_names: Set[str] = {req.project_name for req in module_spec}
+        module_name: str = more_itertools.one(
+            module_names,
+            too_short=ValueError("module_spec should contain at least one requirement"),
+            too_long=ValueError("module_spec should contain requirements for exactly one module"),
+        )
+        return module_name
+
+
+class ModuleV2Source(ModuleSource["ModuleV2"]):
+    def __init__(self, urls: List[str]) -> None:
+        self.urls: List[str] = [url if not os.path.exists(url) else os.path.abspath(url) for url in urls]
```

Comment:
If directory exists the index is a local path, in which case I felt it would be more robust to use the abspath rather than a potentially relative path.